addon/-private/system/record-identifier.ts

```diff
@@ -0,0 +1,238 @@
+import Store from './store';
+import coerceId from './coerce-id';
+import { DEBUG } from '@glimmer/env';
+import { TDict, IResourceIdentifier, IRecordIdentifier } from '../types';
+import { assert } from '@ember/debug';
+
+type TStore = InstanceType<typeof Store>;
+type TmodelName = string;
+type TclientId = string;
+type Tid = string;
+
+interface IKeyOptions {
+  lid: TDict<TclientId, IRecordIdentifier>;
+  id: TDict<Tid, IRecordIdentifier>;
+}
+
+type TTypeMap = TDict<TmodelName, IKeyOptions>;
+
+let CLIENT_ID = 0;
+let DEBUG_MAP;
+
+if (DEBUG) {
+  DEBUG_MAP = new WeakMap<IRecordIdentifier, IRecordIdentifier>();
+}
+
+function generateLid(): string {
+  return `@ember-data:lid-${CLIENT_ID++}`;
```

Comment:
An auto-incrementing integer is not very compatible with offline strategies because it's prone to collisions. If these IDs are all that's used to identify client-side records, then they will need to be persisted to browser storage, and reified along with those records. A new browser instance could not safely restart with `CLIENT_ID = 0`. 

I think a UUID is preferable. A common prefix/namespace is fine though.